xbmc/pvr/providers/PVRProviders.cpp

```diff
@@ -0,0 +1,389 @@
+      if (provider->IsClientProvider())
+        bIgnoreProvider = true;
+
+      if (bIgnoreProvider)
+      {
+        ++it;
+        continue;
+      }
+
+      CLog::LogFC(LOGDEBUG, LOGPVR, "Deleted provider {} on client {}", provider->GetUniqueId(),
+                  provider->GetClientId());
+
+      (*it)->DeleteFromDatabase();
+      it = m_providers.erase(it);
+
+      bChanged = true;
+    }
+    else
+    {
+      ++it;
+    }
+  }
+
+  m_bIsUpdating = false;
+
+  return bChanged;
+}
+
+std::shared_ptr<CPVRProvider> CPVRProviders::CheckAndAddEntry(
+    const std::shared_ptr<CPVRProvider>& newProvider, ProviderUpdateMode updateMode)
+{
+  bool bChanged = false;
+
+  CSingleLock lock(m_critSection);
+  std::shared_ptr<CPVRProvider> provider =
+      GetByClient(newProvider->GetClientId(), newProvider->GetUniqueId());
+  if (provider)
+  {
+    bChanged = provider->UpdateEntry(newProvider, updateMode);
+  }
+  else
+  {
+    // We don't set an id as this came from the DB so it has one already
+    InsertEntry(newProvider, updateMode);
+
+    if (newProvider->GetProviderId() > m_iLastId)
+      m_iLastId = newProvider->GetProviderId();
+
+    provider = newProvider;
+    bChanged = true;
+  }
+
+  if (bChanged)
+    return provider;
+
+  return {};
+}
+
+std::shared_ptr<CPVRProvider> CPVRProviders::CheckAndPersistEntry(
+    const std::shared_ptr<CPVRProvider>& newProvider, ProviderUpdateMode updateMode)
+{
+  bool bChanged = false;
+
+  CSingleLock lock(m_critSection);
+  std::shared_ptr<CPVRProvider> provider =
+      GetByClient(newProvider->GetClientId(), newProvider->GetUniqueId());
+  if (provider)
+  {
+    bChanged = provider->UpdateEntry(newProvider, updateMode);
+
+    if (bChanged)
+      provider->Persist(true);
+
+    CLog::LogFC(LOGDEBUG, LOGPVR, "Updated provider {} on client {}", newProvider->GetUniqueId(),
+                newProvider->GetClientId());
+  }
+  else
+  {
+    newProvider->SetProviderId(++m_iLastId);
+    InsertEntry(newProvider, updateMode);
+
+    newProvider->Persist();
+
+    CLog::LogFC(LOGDEBUG, LOGPVR, "Added provider {} on client {}", newProvider->GetUniqueId(),
+                newProvider->GetClientId());
+
+    provider = newProvider;
+    bChanged = true;
+  }
+
+  if (bChanged)
+    return provider;
+
+  return {};
+}
+
+bool CPVRProviders::PersistUserChanges(const std::vector<std::shared_ptr<CPVRProvider>>& providers)
+{
+  for (auto provider : providers)
+  {
+    provider->Persist(true);
+
+    CLog::LogFC(LOGDEBUG, LOGPVR, "Updated provider {} on client {}", provider->GetUniqueId(),
+                provider->GetClientId());
+  }
+
+  return true;
+}
+
+std::shared_ptr<CPVRProvider> CPVRProviders::GetById(unsigned int iProviderId) const
+{
+  std::shared_ptr<CPVRProvider> item;
+  CSingleLock lock(m_critSection);
+  for (const auto& provider : m_providers)
+  {
+    if (provider->GetProviderId() == iProviderId)
+      return provider;
+  }
+
+  return item;
```

Comment:
return {}; and remove the declaration of item at the beginning of this method. you actually don't use it anymore, then,

Reply:
Done